### ISSUE TITLE
Invalid arguments when running constant_example.py

### ISSUE THREAD

**Issue by jrt54** [opesci/devito#316]
I've installed and setup my environment via conda. Every other example I've tried runs successfully, albeit with occasional warnings about ignoring simd pragmas. 

But I can't get examples/seismic/acoustic/constant_example.py to run at all.  Here's the error code

  Applying Forward
DSE: extract_time_invariants [flops: 44, elapsed: 0.01] >>
     eliminate_inter_stencil_redundancies [flops: 44, elapsed: 0.01] >>
     eliminate_intra_stencil_redundancies [flops: 43, elapsed: 0.02] >>
     factorize [flops: 28, elapsed: 0.03] >>
     finalize [flops: 28, elapsed: 0.00]
     [Total elapsed: 0.08 s]
DLE: analyze [elapsed: 0.05] >>
     avoid_denormals [elapsed: 0.00] >>
     loop_fission [elapsed: 0.02] >>
     loop_blocking [elapsed: 0.23] >>
     simdize [elapsed: 1.28] >>
     create_elemental_functions [elapsed: 0.06] >>
     minimize_remainders [elapsed: 0.01]
     [Total elapsed: 1.64 s]
Traceback (most recent call last):
  File "examples/seismic/acoustic/constant_example.py", line 89, in <module>
    run(full_run=True, space_order=6, time_order=2)
  File "examples/seismic/acoustic/constant_example.py", line 75, in run
    rec, u, summary = solver.forward(save=full_run, dse='advanced', dle='advanced')
  File "/home/mydir/devitotools/devito/examples/seismic/acoustic/wavesolver.py", line 111, in forward
    summary = self.op_fwd_save.apply(src=src, rec=rec, u=u, m=m, **kwargs)
  File "/home/mydir/devitotools/devito/devito/operator.py", line 468, in apply
    arguments, dim_sizes = self.arguments(*args, **kwargs)
  File "/home/mydir/devitotools/devito/devito/operator.py", line 187, in arguments
    raise InvalidArgument("Unknown arguments passed: " + ", ".join(kwargs.keys()))
devito.exceptions.InvalidArgument: Unknown arguments passed: dle, m, dse

**Comment by mloubout** [comment]
Hi @jrt54, yes we found that out an are working on it. Constant parameters changes are still not always handled properly. I would advise to use the normal `acoustic_example.py` with a constant array for the velocity in the mean time.

**Comment by FabioLuporini** [comment]
Thanks for pointing this out and thanks for opening an issue. 

Yes, we were aware of this (https://github.com/opesci/devito/pull/297), but we're still working on a proper fix. We'll keep you posted, hopefully this shouldn't take much

Meanwhile, had you got any more questions, feel free to pop by to our Slack channel (you can get an invitation link at the top right here: http://www.opesci.org/)

**Comment by jrt54** [comment]
I do have a followup to this, as I saw a similar error when working with TTI_3D and assumed they're related (so I didn't want to open a new issue). 

If I change stencilp or stencilr to something simple like

stencilp = u
stencilr = v

I'll get the error posted below. It appears that the stencil must make use of every attribute of the model. I guess I'm not sure if this is a bug or working as intended. There are some applications where I'd like to make use of the same model, but on different PDE's (e.g. isotropic vs TTI vs variable density) and have the stencil defined only by the relevant attributes. Is there a workaround? 

Traceback (most recent call last):
  File "tti_example3D.py", line 77, in <module>
    run()
  File "tti_example3D.py", line 71, in run
    rec, u, v, summary = solver.forward(autotune=autotune)
  File "/home/mydir/devitotools/devito/examples/seismic/tti/wavesolver.py", line 115, in forward
    delta=delta, theta=theta, phi=phi, **kwargs)
  File "/home/mydir/devitotools/devito/devito/operator.py", line 468, in apply
    arguments, dim_sizes = self.arguments(*args, **kwargs)
  File "/home/mydir/devitotools/devito/devito/operator.py", line 187, in arguments
    raise InvalidArgument("Unknown arguments passed: " + ", ".join(kwargs.keys()))
devito.exceptions.InvalidArgument: Unknown arguments passed: epsilon, delta, theta, phi

**Comment by mloubout** [comment]
So this error is (currently) expected. We are working on a fix for the parameter inputs. However a few details will remain.

1- Once a solver is created (tti or acoustic), the types of the physical parametrs are fixed. You are still able to do `solver.forward(m=m_new)` but this will require that, if the original m is an array, `m_new` as to be an array and cannot be a constant. The same stands for the TTI parameters.

2- Concerning you erro, it comes from the fact that the wrapper `forward` calls `operator.apply(epsilon=epsilon, ...)`. However, as your stencil does not contain `epsilon` anymore, the argument persons throws an error. This behavior is expected as Devito will consider that the provided input should be used and that the generated Kernel does not corresponds to the input.

>There are some applications where I'd like to make use of the same model, but on different PDE's

This is possible, (variable density is not included in the examples for now). You will however need to generate a solver for each different PDE, as Devito is based on code generation for a specific problem. 
```
solver_acoustic = AcousticWaveSolver(model, source=source, receiver=receiver,
                                  time_order=2, space_order=space_order)
rec, u, summary = solver.forward()
solver_tti = AnisotropicWaveSolver(model, source=source, receiver=receiver,
                                       time_order=2, space_order=space_order)
rec_tti, u_tti, v_tti, summary = solver.forward()
```

You cloud also create a TTI kernel an then call the operator as (`solver.forward(epsilon = np.ones(solver.model.shape_domain, dtype=np.float32), delta=...)`. However, this would not be usefull for benchmarking as you would compute the solution fo the TTI equations with zeros parameters.

**Comment by FabioLuporini** [comment]
Fixed by #297
